ord_schema/validations.py

```diff
@@ -261,10 +286,39 @@ def validate_dataset(message):
         for reaction_id in message.reaction_ids:
             if not re.fullmatch('^ord-[0-9a-f]{32}$', reaction_id):
                 warnings.warn('Reaction ID is malformed', ValidationError)
-    if message.dataset_id:
+    if validate_id:
         # The dataset_id is a 32-character uuid4 hex string.
         if not re.fullmatch('^ord_dataset-[0-9a-f]{32}$', message.dataset_id):
             warnings.warn('Dataset ID is malformed', ValidationError)
+    # Check cross-references
+    referenced_ids = set()
+    defined_ids = set()
+    for reaction in message.reactions:
+        if reaction.reaction_id:
+            if reaction.reaction_id in defined_ids:
+                warnings.warn(
+                    'Multiple Reactions should never have the same '
+                    'IDs', ValidationError)
+            defined_ids.add(reaction.reaction_id)
+        for reaction_input in reaction.inputs.values():
+            for component in reaction_input.components:
+                for preparation in component.preparations:
+                    if preparation.reaction_id:
+                        referenced_ids.add(preparation.reaction_id)
+                        if preparation.reaction_id == reaction.reaction_id:
+                            warnings.warn(
+                                'A Reaction should not reference '
+                                'its own ID', ValidationError)
+            for crude_component in reaction_input.crude_components:
+                referenced_ids.add(crude_component.reaction_id)
+                if crude_component.reaction_id == reaction.reaction_id:
+                    warnings.warn(
+                        'A Reaction should not reference '
+                        'its own ID', ValidationError)
+    if len(referenced_ids - defined_ids) > 0:
```

Comment:
Prefer to test without `len`; `if referenced_ids - defined_ids` should be enough.